pkg/sfu/rtpmunger.go

```diff
@@ -55,170 +53,157 @@ type SnTs struct {
-		r.snOffsets[r.snOffsetsWritePtr] = r.snOffset
-		r.snOffsetsWritePtr = (r.snOffsetsWritePtr + 1) & SnOffsetCacheMask
-		r.snOffsetsOccupancy++
-	}
-	if r.snOffsetsOccupancy > SnOffsetCacheSize {
-		r.snOffsetsOccupancy = SnOffsetCacheSize
+	// can get duplicate packet due to FEC
+	if diff == 0 {
+		return &TranslationParamsRTP{
+			snOrdering: SequenceNumberOrderingDuplicate,
+		}, ErrDuplicatePacket
 	}
 
 	ordering := SequenceNumberOrderingContiguous
 	if diff > 1 {
 		ordering = SequenceNumberOrderingGap
-	} else {
-		// can get duplicate packet due to FEC
-		if diff == 0 {
-			return &TranslationParamsRTP{
-				snOrdering: SequenceNumberOrderingDuplicate,
-			}, ErrDuplicatePacket
-		}
+		r.snRangeMap.AddRange(r.extHighestIncomingSN+1, extPkt.ExtSequenceNumber)
+	}
 
-		// if padding only packet, can be dropped and sequence number adjusted
-		// as it is contiguous and in order. That means this is the highest
-		// incoming sequence number, and it is a good point to adjust
-		// sequence number offset.
-		if len(extPkt.Packet.Payload) == 0 {
-			r.highestIncomingSN = extPkt.Packet.SequenceNumber
-			r.snOffset++
+	r.extHighestIncomingSN = extPkt.ExtSequenceNumber
 
-			return &TranslationParamsRTP{
-				snOrdering: SequenceNumberOrderingContiguous,
-			}, ErrPaddingOnlyPacket
-		}
+	// if padding only packet, can be dropped and sequence number adjusted, if contiguous
+	if diff == 1 && len(extPkt.Packet.Payload) == 0 {
+		r.snRangeMap.IncValue(1)
+		return &TranslationParamsRTP{
+			snOrdering: ordering,
+		}, ErrPaddingOnlyPacket
 	}
 
-	// in-order incoming packet, may or may not be contiguous.
-	// In the case of loss (i.e. incoming sequence number is not contiguous),
-	// forward even if it is a padding only packet. With temporal scalability,
-	// it is unclear if the current packet should be dropped if it is not
-	// contiguous. Hence, forward anything that is not contiguous.
-	// Reference: http://www.rtcbits.com/2017/04/howto-implement-temporal-scalability.html
-	mungedSN := extPkt.Packet.SequenceNumber - r.snOffset
+	snOffset, err := r.snRangeMap.GetValue(extPkt.ExtSequenceNumber)
+	if err != nil {
+		return &TranslationParamsRTP{
+			snOrdering: ordering,
+		}, ErrSequenceNumberOffsetNotFound
+	}
+
+	extMungedSN := extPkt.ExtSequenceNumber - snOffset
 	mungedTS := extPkt.Packet.Timestamp - r.tsOffset
 
-	r.highestIncomingSN = extPkt.Packet.SequenceNumber
-	r.lastSN = mungedSN
+	r.extLastSN = extMungedSN
 	r.lastTS = mungedTS
 	r.lastMarker = extPkt.Packet.Marker
 
 	if extPkt.KeyFrame {
-		r.rtxGateSn = mungedSN
+		r.extRtxGateSn = extMungedSN
 		r.isInRtxGateRegion = true
 	}
 
-	if r.isInRtxGateRegion && (mungedSN-r.rtxGateSn) < (1<<15) && (mungedSN-r.rtxGateSn) > RtxGateWindow {
+	if r.isInRtxGateRegion && (extMungedSN-r.extRtxGateSn) > RtxGateWindow {
 		r.isInRtxGateRegion = false
 	}
 
 	return &TranslationParamsRTP{
 		snOrdering:     ordering,
-		sequenceNumber: mungedSN,
+		sequenceNumber: uint16(extMungedSN),
```

Comment:
Made a comment in PR commit notes about keeping extended types. An example here, i. e. would be good to return extended type and when this value actually gets used to construct an RTP packet, it can be truncated to 16-bits.